sstables/generation_type.hh

```diff
@@ -36,6 +37,33 @@ constexpr int64_t generation_value(generation_type generation) {
     return generation.value();
 }
 
+struct is_shared {
+    using yes = std::true_type;
+    using no = std::false_type;
+};
+
+template<typename IsShared>
+class sstable_generation_generator {
+    // We still want to do our best to keep the generation numbers shard-friendly.
+    // Each destination shard will manage its own generation counter.
+    //
+    // operator() is called by multiple shards in parallel when performing reshard,
+    // so we have to use atomic<> here.
+    std::conditional_t<IsShared::value, std::atomic<int64_t>, int64_t> _last_generation;
+public:
+    explicit sstable_generation_generator(int64_t last_generation)
+        : _last_generation(last_generation) {}
+    sstables::generation_type operator()(shard_id shard) {
+        int64_t v;
+        if constexpr (IsShared::value) {
+            v = _last_generation.fetch_add(smp::count, std::memory_order_relaxed);
+        } else {
+            v = (_last_generation += smp::count);
+        }
+        return sstables::generation_from_value(v + shard);
+    }
+};
+
```

Comment:
> what's your plan when extending it to UUID?
> 
> IsShared template parameter will be bogus for UUID, right?
> 

yeah, the `IsShared` setting is a compile-time configuration, while UUID/integer is a runtime setting. the implementation would look like https://github.com/tchaikov/scylladb/commit/c50d7bf98e099a1f6bcdf901acbca7c6bd230572#diff-2019b4ff0737c2b01c33f32fbdb5e54fa68b98af76250fc7307e58ff2de5b613R22

> I am thinking if it it wouldnt be simpler to just stick to std::atomic, as the frequency of generator being called correlates with a new sstable creation, and that happens only once in >1 seconds. The difference for simple arithmetic is in the order of very few microseconds, so the overhead due to atomicity guarantees won't really affect the overall performance
> 

yeah, good point indeed. but i think it'd be better to be more explicit, as we are working on a project which enforces a share-nothing design. whenever one sees `atomic<>`, he/she would feel uncomfortable and starts asking "why?". and personally, i don't think it matters much. the only place we need to use a "shared" generator is when we do reshard. once the sstables find their home and are settled, we are back to a share-nothing world.

> As a benefit, transition to UUID is easier without the template parameter.

we could either have a default template parameter or define an alias for the more popular one -- the non-shared one.

Reply:
> we could either have a default template parameter or define an alias for the more popular one -- the non-shared one.

Or instantiate a `sharded<>` one where needed and use invoke_on instead of using shared std::atomic state.

Reply:
@bhalevy we already have a sharded generator in effect, the problem is that callers call from across shards.

Reply:
Why not put the generator in `generation_type.hh` in the first place?